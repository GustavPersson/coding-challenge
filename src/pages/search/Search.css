.search-button {
    background-color: "#3f51b5";
    color: "#fff";
}

.search-button:hover {
    background-color: "#3f51b5";
    color: "#fff";
  };