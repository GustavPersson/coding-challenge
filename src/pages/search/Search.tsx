import {
  Avatar,
  Box,
  Button,
  Grid,
  List,
  ListItem,
  ListItemAvatar,
  ListItemText,
  Stack,
  TextField,
  Typography,
  Link,
} from "@mui/material";
import { Observer } from "mobx-react-lite";
import { ChangeEventHandler, ReactElement, useEffect, useState } from "react";
import ImageIcon from "@mui/icons-material/Image";
import { SearchType } from "./SearchViewModel";
import { action } from "mobx";
import { Link as ReactRouterLink, useSearchParams } from "react-router-dom";
import "./Search.css";
import useContainer from "../../utils/useContainer";

function SearchResultsPage({
  setTitle,
}: {
  setTitle: (title: string) => void;
}): ReactElement {
  const viewModel = useContainer<SearchType>("SEARCH_PAGE");
  const [searchParams, setSearchParams] = useSearchParams();
  const [searchValue, setSearchValue] = useState(
    searchParams.get("query") || ""
  );
  const [isSearchPristine, setIsSearchPristine] = useState(true);

  const onChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    setSearchValue(event.target.value);
  };

  useEffect(() => {
    setTitle("Search");
  }, []);

  useEffect(() => {
    if (searchParams.get("query")) {
      viewModel.search(searchValue);
      setIsSearchPristine(false);
    }
  }, []);

  return (
    <Observer>
      {() => (
        <Box>
          <Grid container spacing={2}>
            <Grid item md={4}>
              <Stack direction="row">
                <TextField
                  id="outlined-basic"
                  label="Search for an article"
                  variant="outlined"
                  value={searchValue}
                  onChange={onChange}
                />
                <Button
                  data-testid="search"
                  className="search-button"
                  sx={{ marginLeft: 2 }}
                  variant="contained"
                  onClick={action(() => {
                    setSearchParams({ query: searchValue });
                    viewModel.search(searchValue);
                    setIsSearchPristine(false);
                  })}
                >
                  Search
                </Button>
              </Stack>
            </Grid>
            <Grid item md={8}>
              <Box alignItems="center" flexGrow={1}>
                <Typography align="center" variant="h5">
                  Search results
                </Typography>
              </Box>
              {!isSearchPristine &&
                searchValue &&
                viewModel.searchResults.length === 0 && <Box>No results</Box>}
              <List>
                {viewModel.searchResults &&
                  viewModel.searchResults.map((result) => (
                    <Link
                      data-testid="article-link"
                      component={ReactRouterLink}
                      to={`/article/${result.id}`}
                      key={result.id}
                    >
                      <ListItem
                        onClick={() => {
                          console.log(result.id);
                        }}
                      >
                        <ListItemAvatar>
                          <Avatar>
                            <ImageIcon />
                          </Avatar>
                        </ListItemAvatar>
                        <ListItemText
                          primary={result.title}
                          secondary={result.description}
                        />
                      </ListItem>
                    </Link>
                  ))}
              </List>
            </Grid>
          </Grid>
        </Box>
      )}
    </Observer>
  );
}

export default SearchResultsPage;
