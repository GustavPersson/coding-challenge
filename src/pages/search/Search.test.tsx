import { render, waitFor, screen } from "@testing-library/react";
import Search from "./Search";
import {
  MemoryRouter,
  RouterProvider,
  createMemoryRouter,
} from "react-router-dom";

describe("SearchPage", () => {
  const setTitle = jest.fn();
  const routes = [
    {
      path: "/search",
      element: <Search setTitle={setTitle} />,
    },
  ];
  const router = createMemoryRouter(routes, {
    initialEntries: ["/search"],
  });

  beforeEach(() => {
    setTitle.mockReset();
  });

  it("should render", async () => {
    const { container } = render(<RouterProvider router={router} />);
    await waitFor(() => screen.getByText("Search"));
    expect(container).toMatchSnapshot();
    expect(setTitle).toBeCalledWith("Search");
  });

  it("should render with results", async () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/search?query=iphone"]}>
        <Search setTitle={setTitle} />
      </MemoryRouter>
    );
    await waitFor(() => screen.getByText("iPhone 9"));

    expect(container).toMatchSnapshot();
    expect(setTitle).toBeCalledWith("Search");
  });
});
