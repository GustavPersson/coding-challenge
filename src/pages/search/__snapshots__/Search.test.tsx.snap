// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage should render 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4 css-1ube6tp-MuiGrid-root"
      >
        <div
          class="css-m69qwo-MuiStack-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="outlined-basic"
              id="outlined-basic-label"
            >
              Search for an article
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="outlined-basic"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Search for an article
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium search-button css-1lddlii-MuiButtonBase-root-MuiButton-root"
            data-testid="search"
            tabindex="0"
            type="button"
          >
            Search
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-8 css-am41oj-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-1oaa9v5"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
          >
            Search results
          </h5>
        </div>
        <ul
          class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchPage should render with results 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4 css-1ube6tp-MuiGrid-root"
      >
        <div
          class="css-m69qwo-MuiStack-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="outlined-basic"
              id="outlined-basic-label"
            >
              Search for an article
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="outlined-basic"
                type="text"
                value="iphone"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Search for an article
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium search-button css-1lddlii-MuiButtonBase-root-MuiButton-root"
            data-testid="search"
            tabindex="0"
            type="button"
          >
            Search
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-8 css-am41oj-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-1oaa9v5"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
          >
            Search results
          </h5>
        </div>
        <ul
          class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
            data-testid="article-link"
            href="/article/1"
          >
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
            >
              <div
                class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ImageIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21 19V5c0-1.1-.9-2-2-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zM8.5 13.5l2.5 3.01L14.5 12l4.5 6H5l3.5-4.5z"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                >
                  iPhone 9
                </span>
                <p
                  class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                >
                  An apple mobile which is nothing like apple
                </p>
              </div>
            </li>
          </a>
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
            data-testid="article-link"
            href="/article/2"
          >
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
            >
              <div
                class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ImageIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21 19V5c0-1.1-.9-2-2-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zM8.5 13.5l2.5 3.01L14.5 12l4.5 6H5l3.5-4.5z"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                >
                  iPhone X
                </span>
                <p
                  class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                >
                  SIM-Free, Model A19211 6.5-inch Super Retina HD display with OLED technology A12 Bionic chip with ...
                </p>
              </div>
            </li>
          </a>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
